addons/account/models/account_move_line.py

```diff
@@ -2949,13 +2939,22 @@ def _reconciled_lines(self):
 
     def _all_reconciled_lines(self):
         reconciliation_lines = self.filtered(lambda x: x.account_id.reconcile or x.account_id.account_type in ('asset_cash', 'liability_credit_card'))
-        current_lines = reconciliation_lines
-        current_partials = self.env['account.partial.reconcile']
-        while current_lines:
-            current_partials = (current_lines.matched_debit_ids + current_lines.matched_credit_ids) - current_partials
-            current_lines = (current_partials.debit_move_id + current_partials.credit_move_id) - current_lines
-            reconciliation_lines += current_lines
-        return reconciliation_lines
+        self.env['account.partial.reconcile'].flush_model()
+        self.env.cr.execute("""
+            WITH RECURSIVE partials (current_id) AS (
+                    SELECT line.id
+                      FROM account_move_line line
+                     WHERE id = ANY(%s)
+
+                                                UNION
+
+                    SELECT CASE WHEN partial.debit_move_id = p.current_id THEN partial.credit_move_id ELSE partial.debit_move_id END
+                      FROM partials p
+                      JOIN account_partial_reconcile partial ON partial.debit_move_id = p.current_id OR partial.credit_move_id = p.current_id
+            )
+            SELECT current_id FROM partials;
+        """, [reconciliation_lines.ids])
+        return self.browse(r[0] for r in self.env.cr.fetchall())
```

Comment:
This could actually be
```suggestion
        return self.sudo().search([('matching_number', 'in', set(reconciliation_lines.mapped('matching_number'))])
```

Reply:
@william-andre This is very dangerous. If your implementation has a single bug, you could break the whole reconciliation :grimacing: